RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,132 @@
+## Requirements of `FullyQualifiedName` and TestCase `ID`
 
-| Fully Qualified Name | Normalized Fully Qualified Name |
-|---|---|
-| `Namespace.Namespace.Class.Method()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<Int32>.Method()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<Int32>.Method<String>()` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class+InnerClass.Method` | `Namespace.Namespace.Class+InnerClass.Method` |
-| `Namespace.Namespace.Class.Method(x: 1, msg: "string")` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class.Method(1, "string")` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class(init: "Value").Method(param: 23)` | `Namespace.Namespace.Class.Method` |
-| `Namespace.Namespace.Class<String>("Value").Method(23)` | `Namespace.Namespace.Class.Method` |
+* `FullyQualifiedName` should uniquely identify a method that represents test cases (with the optional application of a dataset). There may be duplicate FQNs within a project due to some methods being associated with multiple test cases. Note that it is possible for multiple assemblies, or multiple test executors to use the same FQNs. When comparing FQNs within a Visual Studio session, `Source`, `FQN`, and `ExecutorURI` must be combined to guarantee uniqueness.
+* Each TestCase has an `ID` field that takes the form of a GUID. In the past, this field has been optional. It will now become required. The required `ID` field must be unique for a given test assembly. This `ID` field must also be deterministic, in the sense that a test with the same class, method and method arguments should have the same id. We envision generating this `ID` by hashing the test's class, method, method signature and method arguments into the GUID. 
+* `FullyQualifiedName` should be immutable in that the test store keeps the exact value that is provided by the test adapter and supplies it back to the test adapter for test execution. The test adapter should be able to rely on receiving the same FQN for execution that it supplied when performing discovery.
+* `FullyQualifiedName` should reference the semantic location of the method that defines the test. This means that the `FullyQualifiedName` references the method on the type where the test is discovered. In some situations, this could be different from the type where the method is defined. See later in this document for a more detailed discussion.
 
+## `FullyQualifiedName` specification
 
-The normalized form allows the test platform to reference groups of tests in a more general form without having to interpret the logic that the test adapter uses when constructing the fully qualified name. 
+We are standardizing the `FullyQualifiedName` property on a TestCase in the following way. The FQN will be parsed as a URI starting with the scheme value of `fqn`. 
+
+**Method Form**
+
+    fqn://clr/m/{type}/{method}
```

Comment:
There is a Test Plan/Test Suite scenario in vsts which requires [association](https://docs.microsoft.com/en-us/vsts/test/associate-automated-test-with-test-case?view=vsts) of test method. Associations stores FQN.

1. Currently we support association for mstest v1 adapter only. Changing the FQN for mstest v1 adapter will break all these existing TestPlan/TestSuite scenarios.
2. In addition, there is a huge ask from community to enable association for all adapters (including mstest v2, XUnit, NUnit). Once association is enabled for mstest v2 adapter, all the existing TestPlan scenarios which points to mstestv1 will not be able to migrate to mstest v2 (if this FQN change happens in mstest v2).

Reply:
There is nothing wrong with storing the standardized version of the FQN. It is just a string that identifies a test case.

An adapter can tell the difference between an existing FQN and the standardized version by looking to see if the string starts with `fqn://`. So, it would know whether the FQN is standardized or not and would be able to adjust as needed. We also have a standard for representing the older FQN as a standardized FQN by the 'other' form (i.e. `fqn://clr/o/{non-standard fqn}`). So, if necessary, the non-standard FQN could be up-converted to a standardized form without interpreting it.

Reply:
'Legacy compatibility' section of the PR mentions that how TestWindow will handle the scenario when adapters does not specify the proposed standardized FQN. 
It doesn't give clarity that adapter also needs to convert old FQN to standardized format.

So apart from moving to the new standard FQN, is there an expectation from adapters to have a compat mechanism as well (in which adapters converts the old FQN to new FQN)? Can you please update the RFC if that's the case?

This compat conversion seems more like the job of TestWindow or TestPlatform. Adapter should not be doing such compat conversions.

Reply:
No, we don't expect adapters to need to convert the old FQN to the new FQN. We are only asking that the adapter switch to the new format. The runsetting will be provided to allow for handling compat issues, where the adapter would like to be aware if Std FQN is supported.